### PR TITLE
git diff support

### LINKED ISSUE #295: act as textconv so can git diff inventory
**Is your feature request related to a problem? Please describe.**
git can be informed on which files are binary. And then instructed how to convert the binary files into text files, so ```git diff``` can work. 

git passes in the .inv file name cuz the ``git diff`` is occurring on the .inv, not on the .txt file.

```sphobjinv convert plain``` requires a hyphen to be passed in to have the output sent to stdout. Instead would like a option flag. So then sphobjinv can act as a textconv

docs/.gitattributes

```
[diff "inv"]
  textconv = sphobjinv --stdout convert plain 
  binary = true
```

**Describe the solution you'd like**
Provide a --stdout option flag as another means to specify the outfile hyphen.

**Describe alternatives you've considered**
Alternative approach, which i'm using now, is to mark the inventory files as binary and not be able to use ```git diff```

docs/.gitattributes

```
*.inv binary
```

**Additional context**

#159 is a similar feature request, but for stdin
The difference is infile is a required positional argument whereas outfile is optional. So a --stdout flag would not change sphobjinv command line interface

[perform diffs on binary files](https://git-scm.com/docs/gitattributes#_performing_text_diffs_of_binary_files)

the fallback, [indicate files are binary](https://git-scm.com/docs/gitattributes#_using_macro_attributes)

### PR DESCRIPTION
**Pull request by msftcangoblowm**
- feat: entrypoint sphobjinv-textconv #295 
- test: add sphobjinv-textconv unittests offline and online
- test: add integration test. Demonstrate git diff objects.inv
- test: add pytest module with class to interact with git
- docs: add step by step guide to configure git. Both bash and python code examples
- docs: added sphobjinv-textconv API docs

**Is the PR a fix or a feature?**
This is a new feature release

Bump the version up to 2.4.0

**Describe the changes in the PR**
Closes #295 

**Does this PR close any issues?**
Closes #295 

**Does the PR change/update the following, if relevant?**

- [X] Documentation
- [X] Tests
- [X] CHANGELOG
- [X] version bump
- [X] doctest
- [X] checklink
- [X] tox -e flake8-noqa
- [X] tox -e interrogate
- [X] pre-commit
- [X] Integration tests
- [X] 100% coverage
- [no] static type checking
- [no] modifies sphobjinv entrypoint codebase
- [no] ran CI/CD workflows (can't)

### CODE REVIEW

**Comment by msftcangoblowm** [comment]
**I'm flying blind**

Not a Windows user nor have access to a Windows box or VM

This is a plea for help!

On Windows, how to get the absolute path of these executables: sphobjinv and sphobjinv-textconv

These are incorrect

```
path_soi = Path(sys.executable).parent.joinpath("sphobjinv")
path_soi_textconv = Path(sys.executable).parent.joinpath("sphobjinv-textconv")
```

Affects

- tests/test_cli_textconv
- tests/test_cli_textconv_with_git

**Comment by msftcangoblowm** [comment]
Packages site path: C:\Users\VssAdministrator\AppData\Roaming\Python\Python38\site-packages

site packages: ['c:\\hostedtoolcache\\windows\\python\\3.8.10\\x64', 'c:\\hostedtoolcache\\windows\\python\\3.8.10\\x64\\lib\\site-packages']

user site packages: 'C:\\Users\\VssAdministrator\\AppData\\Roaming\\Python\\Python38\\site-packages'

path_scripts: C:\Users\VssAdministrator\AppData\Roaming\Python\Python38\SCRIPTS

**Comment by bskinn** [comment]
I would say, take a look at the other CLI tests and see if it's possible to adapt the approach I took with them to your new tests.

It's possible the needs here are different enough that my existing technique is insufficient, but I'd at least give it a shot first.

(That is, if you haven't already.)

**Comment by bskinn** [comment]
I'm not sure where the script is; looking at the output of [this job step], though, it looks like it's in `D:\a\1\s`, at least for this run.

You could try the `where` command -- it's analogous to `which` on Linux.


[this job step]: https://dev.azure.com/brianskinn/sphobjinv/_build/results?buildId=513&view=logs&jobId=489e200e-1db0-5fe5-8212-912b67c676b3&j=489e200e-1db0-5fe5-8212-912b67c676b3&t=d2033042-e5cb-5c6b-0f43-edbcaea58dd7

**Comment by bskinn** [comment]
I'm also not sure why you need to invoke the `sphobjinv-textconv` with the full path to the entrypoint script. It should be on path; can't you just call it without any path specified?

**Comment by msftcangoblowm** [comment]
Will follow your advice and try with a relative path rather than an absolute path. You are right, the venv bin|SCRIPTS folder should already be on the path

File: tests/test_cli_textconv.py [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+            #
+            # From an UX POV, running echo $? and getting 1 on error is
+            # useless and frustrating.
+            #
+            # Not relaying errors and giving exact feedback on how to rectify
+            # the issue is bad UX.
+            #
+            # So if the numerous exit codes of 1 looks strange. It is; but this is
+            # a separate issue best solved within a dedicated commit
+            assert f"EXIT CODES{os.linesep}" in str_out
```
Comment (bskinn) [inline tests/test_cli_textconv.py:103]: I'll definitely be taking a closer look at this, for the project as a whole. I appreciate you pointing it out.

Could you actually duplicate these thoughts into a new ticket on the project?

(I'm aware that the exception handling in the project is substandard and intend to do something about it eventually (see #118), but it hadn't occurred to me to tie an exception hierarchy directly to the CLI exit codes in this way.)

File: CHANGELOG.md [hunk 2]
```diff
@@ -8,6 +8,101 @@ and this project follows an extension of
 fourth number represents an administrative maintenance release with no code
 changes.
 
+### [2.4.1.10] - 2024-08-28
```
[the remark this answers is not in the document]
Comment (msftcangoblowm) [reply CHANGELOG.md:11]: This is fine.

These are actually development prereleases, not administrative prereleases. So these entries are wrong and spammy. The final changelog entries are left to your discretion

Also refrained from making any tags

File: tests/test_cli_textconv_with_git.py [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+        path_git_config.write_text(str_git_config)
+        if platform.system() in ("Linux", "Windows"):
+            print(f"cwd {wd.cwd!s}", file=sys.stderr)
+            print(f"executable path: {resolved_soi_textconv_path}", file=sys.stderr)
+            print(f"""PATHEXT: {os.environ.get("PATHEXT", None)}""", file=sys.stderr)
+            print(f".git/config {str_git_config}", file=sys.stderr)
+
+        #    .gitattributes
+        #    Informs git: .inv are binary files and which cmd converts .inv --> .txt
+        path_ga = path_cwd / ".gitattributes"
```
[the remark this answers is not in the document]
Comment (msftcangoblowm) [reply tests/test_cli_textconv_with_git.py:142]: The cwd is a temp folder. ```git init``` does not create a .gitattributes

In the temp folder, the file does not exist until it's created.

The source of confusion, is this is written for the future. When the code would be encapsulated and an entrypoint created to greatly improve UX. Has not been done, cuz left for a future PR (outside of scope).

Initializing (.git/config), on Windows is tricky. Requires an absolute path to the executable, not a relative path. In the future, would update these config files, not overwrite them.

File: tests/test_cli_textconv_with_git.py [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+        str_gitattributes = f"{str_gitattributes}{sep}{ga_textconv}"
+        wd.write(".gitattributes", str_gitattributes)
+
+        #    commit (1st)
+        wd.add_command = "git add ."
+        wd.commit_command = "git commit --no-verify --no-gpg-sign -m test-{reason}"
+        wd.add_and_commit(reason="sphobjinv-textconv", signed=False)
+
+        # Act
+        #    make change to .txt inventory (path_dst_dec)
```
[the remark this answers is not in the document]
Comment (msftcangoblowm) [reply tests/test_cli_textconv_with_git.py:155]: This is a realistic workflow. A line is added to a .txt --> soi --> .inv

``soi convert -q zlib dec cmp`` with resource absolute paths did not work on Windows. So that issue would not have been found, if had not gone thru the entire workflow.

Will isolate that issue into a separate test and then follow your advice on using two pre-built resources

[the remark this answers is not in the document]
Comment (bskinn) [reply tests/test_cli_textconv_with_git.py:155]: Never mind on this -- I was thinking about it incorrectly. `git diff` acts on the same file over time, not on two different files.

Instead, I'd recommend creating a small, trivial `objects_textconv.inv` for this, and then making a new version and committing a new version of that same `objects_textconv.inv`. That will give the two different versions of the files to be `git diff`-ed.

And then, the `git diff` command in the test would be something like `git diff <sha1>..<sha2> tests/resource/objects_textconv.inv`.

I might still not have this quite right, but it's at least closer.

[the remark this answers is not in the document]
Comment (bskinn) [reply tests/test_cli_textconv_with_git.py:155]: > `soi convert -q zlib dec cmp` with resource absolute paths did not work on Windows. So that issue would not have been found, if had not gone thru the entire workflow.

This doesn't make sense to me, I can use absolute paths on Windows without problems in normal operation mode.

It could be something going funny with the Azure CI environment, but it's concerning that absolute paths aren't working correctly.

[the remark this answers is not in the document]
Comment (msftcangoblowm) [reply tests/test_cli_textconv_with_git.py:155]: In the next commit, followed your advice on fixtures and created

- `tests/resource/objects_attrs_plus_one_entry.inv`
- `tests/resource/objects_attrs_plus_one_entry.txt`

`tests/resource/objects_attrs_plus_one_entry.inv` overwrites `cwd/objects.inv`

File: tests/wd_wrapper.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+import itertools
+import os
+import shlex
+import subprocess as sp  # noqa: S404
+from pathlib import Path
+from typing import List
+
+
+def run(
```
[the remark this answers is not in the document]
Comment (msftcangoblowm) [reply tests/wd_wrapper.py:21]: `tests/wd_wrapper.py` is a standalone component under setuptools-scm author's copyright. (my voice not the setuptools-scm author) It should be fully functional by itself, without dependencies outside of StdLib

The (subprocess wrapper module func) run, i provided, replaces the one offered by `setuptools-scm` in the original `tests/wd_wrapper.py`. Adding a `setuptools-scm` dependency for just one function in one module maybe a bad idea.

Elsewhere in the test suite, `subprocess.check_output` is used. It's obsoleted by `subprocess.run`.

sphobjinx has no dedicated `_run_cmd` module

In `test_cli_textconv.py`, `TestTextconvStdioFail.test_cli_textconv_zlib_inv_stdin` uses param subprocess.run.input to pipe in stdin and do a subprocess in one command! It's a little exotic, non-typical usage. So different from what's normally in a `_run_cmd` module.

[_run_cmd](https://github.com/pypa/setuptools-scm/blob/d665621e25f091c0de5cd35b3d57cc00bd26e260/src/setuptools_scm/_run_cmd.py)

[wd_wrapper.py original](https://github.com/pypa/setuptools-scm/blob/d665621e25f091c0de5cd35b3d57cc00bd26e260/testing/wd_wrapper.py)

File: tests/test_cli_textconv.py (hunk 1 above)
Comment (msftcangoblowm) [reply tests/test_cli_textconv.py:103]: Thanks for accepting the exception handling in cli entrypoints as a worthy issue.

Will create a new issue(s) ... after completing #295

File: tests/test_cli_textconv_with_git.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (msftcangoblowm) [reply tests/test_cli_textconv_with_git.py]: This could be a separate PR. Confirm, on Windows and Azure CI, what's the deal with absolute resource paths and soi convert

In this PR, that code has been removed

**Comment by bskinn** [comment]
I did some digging and testing -- a separate `sphobjinv-textconv` entrypoint that takes the single path argument isn't actually needed.

Combining [this Stack Overflow answer][so] with [this section of the Git docs][git], if I add the following to my `.gitattributes` file:

```none
*.inv diff=objects_inv
```

And add this to my `.git/config`:

```none
[diff "objects_inv"]
	textconv = sh -c 'sphobjinv co plain "$0" -'
```

It works:

```none
(sphobjinv) C:\git\sphobjinv>git diff
diff --git a/.gitattributes b/.gitattributes
index c6f20db..2a23375 100644
--- a/.gitattributes
+++ b/.gitattributes
@@ -1,2 +1,3 @@
 tests/resource/objects_mkdoc_zlib0.inv binary
 tests/resource/objects_attrs.txt binary
+*.inv diff=objects_inv




diff --git a/tests/resource/objects_textconv.inv b/tests/resource/objects_textconv.inv
index 31805f5..19a4323 100644
--- a/tests/resource/objects_textconv.inv
+++ b/tests/resource/objects_textconv.inv
@@ -2,6 +2,6 @@
 # Project: textconv-test
 # Version: 2.3
 # The remainder of this file is compressed using zlib.
-sphobjinv.cli.convert py:module 0 cli/implementation/convert.html#module-$ -
-sphobjinv.cli.convert.do_convert py:function 1 cli/implementation/convert.html#$ -
+sphobjinv.cli.convert23 py:module 0 cli/implementation/convert.html#module-$ -^M
+sphobjinv.cli.convert.do_convert234 py:function 1 cli/implementation/convert.html#$ -^M
```

So, while I really appreciate all of the effort you've gone to here, I believe you can already do what you want with the above recipe.

Open to correction, though, if I'm missing something.

[git]: https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes#_diffing_binary_files
[so]: https://stackoverflow.com/a/58041884/4376000

**Comment by msftcangoblowm** [comment]
Editing the file manually is easy, via code nontrivial.

- On Windows, `sphobjinv` executables path must be resolved

- On Windows, inventory path must be surrounded by single quotes so the backslashes are not removed by bash resulting in:
`C:hostedtoolcachewindowsPython3.8.10x64Scriptssphobjinv-textconv.EXE`

- On Windows, double quotes does not do the backslash escaping. No idea, if "$0" would work as expected

```
git config diff.inv.textconv "sh -c 'sphobjinv co plain \"\$0\" -'"
git config diff.inv.textconv
```

```sh -c 'sphobjinv co plain "$0" -'```

Bash quote escaping is a nightmare, best avoided. UX wise, this is a nasty ugly non-obvious non-trivial solution requiring advanced bash skills and substantial hair loss (stressful).

End of the day, the end user needs smooth UX. These details at this level of complexity would have to be hidden.

**Comment by bskinn** [comment]
I used my above-noted setup successfully without modification on both Windows (with Git for Windows) and Debian WSL, so it definitely works across those platforms.

---

What is the goal of editing the files via code? Providing the setup machinery so a user can run a command and the git textconv is set up automatically for them?

I'm confident (but still possibly wrong!) that the sphobjinv-textconv entry point, and thus also all the new testing around it, are not actually needed.

I like this capability a lot, but I don't want to add an 'automatically set up inv textconv' functionality to sphobjinv. It seems like it would be a bottomless source of support requests and edge case fixes.

It seems to me this is best handled by documenting and publicizing the method for setting it up. Different users will want it set up differently, and so the knowledge will be more valuable than automation here anyways.

**Comment by msftcangoblowm** [comment]
1. A simple python solution is always better than a complicated bash solution. Even if you can, doesn't mean you should

2. UX wise, either method, but especially the bash method, needs a setup entrypoint. Could add that to soi-textconv. Didn't, to keep the PR scope limited. ```sphobjinv-textconv --setup```. The bash route, recommend not adding to soi entrypoint. Already complicated.

So it's an UX issue. One command ... done. Much nicer than having to explain the intricacies of a complex sewer system. 

The interface is simple and easy to understand.

Or we can have super complicated. Requiring an education and long explanation.

Explain both to your wife or a six year old. Then let them cast the deciding vote

**Comment by msftcangoblowm** [comment]
> What is the goal of editing the files via code?

Simplicity. Zero learning curve. Doing less work. Type one simple command ... done.

> Seems like it would be a bottomless source of support requests and edge case fixes

Internally, soi-textconv runs ```git config```. No one would have to manually edit the ```.git/config``` nor the ```.gitattributes``` files or care about the details.

Besides small changes to the parser, all the code for the setup has already been written and is known to work. Just refactor the integration unittest. After a refactor, testing would be similar to that of ```--version``` or ```--help``` 

The scope of soi-textconv is it configures git to understand inventory files. That scope shouldn't change besides the ```--setup``` and maybe ```--global``` flags.

**Comment by bskinn** [comment]
@msftcangoblowm 

TL;DR - You've done a lot of work here, but I'm afraid it's too much... more than I'm willing to add to the project. There's still an avenue to a smaller contribution, though.

---

I'm on board with:

- A new `sphobjinv-textconv` entrypoint to allow the git config entry to be cleaner
- A bit of basic documentation of how to set up the `.inv` textconv
  - Your current location for it in the docstring for the new `cli/core_textconv.py` makes sense for now, but I might change my mind
- A couple of targeted tests to exercise the new entrypoint.

But I don't want to uptake anything more than that:

- I would want the new entrypoint to be narrowly scoped.
  - It should take no options, and only the single `INFILE` argument.
  - The main `sphobjinv` entrypoint should remain the primary, and only configurable, CLI interface surface.
- For tests:
  - No direct testing of the operation of the git textconv.
  - Nothing new in the test suite driving git or editing `.gitattributes` or `.git/config`.
- For docs:
  - No tutorials on anything git other than brief instructions for setting up the textconv. It's just beyond the scope for the `sphobjinv` documentation.

So—if you're interested in paring down the PR to just the above (or, starting a new PR to only implement the above, which might be easier/faster), I think there's still a path to a merge here. Let me know if you want to move forward, and I'll leave more thorough comments/feedback on the PR.

I apologize that you went to all this work before getting this feedback, I didn't have any sense before you started that you were envisioning something of this scale. Lesson learned, I should've started a conversation about scale and approach before you started working!

---

All of this said, I'm still not convinced that the new entrypoint is actually required. Using user-scope `.gitconfig` and `attributes` files, and installing `sphobjinv` with `pipx`, I've been able to get the textconv working on both Git for Windows on Win11 and on Debian WSL without having a `sphobjinv`-containing virtualenv activated.

In `$HOME/.gitconfig` (WSL) & `%USERPROFILE%/.gitconfig` (Win11):

```none
[diff "objects_inv"]
  textconv = sh -c 'sphobjinv co plain "$0" -''
```

In `$HOME/.config/git/attributes` (WSL) & `%USERPROFILE%/.config/git/attributes` (Win11):

```none
*.inv diff=objects_inv
```

For WSL (similar but nonidentical on Win11):

```bash
$ sudo apt install -y pipx
$ pipx install sphobjinv
```

As we've already discussed, the benefit of `sphobjinv-textconv` would be a simpler invocation in the `.gitconfig`... but not *that* much simpler, given that either way, it's two canned snips of config for users to add into the right files.

It's really not much more than a quick post on a blog or Medium or wherever.

(I do understand that a magic command to insert those snips _for_ users would make it even simpler for those users; but, that magic command would end up needing to accommodate a wide variety of operating systems and user setups, and I'm _confident_ it would run into all sorts of hairy edge cases that would be a nightmare to debug and develop around. Far better IMO just to educate users on how to set it up, rather than trying to write something that will attempt to automatically set it up for a wide range of users.)

Regardless, again, if you're still interested in putting together the smaller-scope contribution for adding `sphobjinv-textconv`, I'm on board with working with you toward that.

**Comment by msftcangoblowm** [comment]
> The main sphobjinv entrypoint should remain the primary, and only configurable, CLI interface surface.

If you feel strongly `sphobjinv --setup` is better UX than `sphobjinv-textconv --setup`, i'm totally on board with whatever results in better UX.

This is actually a good idea that did not occur to me.

If the flag should have another name, just suggest a better flag name.

> It should take no options, and only the single INFILE argument.

The `--setup` flag has to go somewhere. If it goes onto the sphobjinv CLI, then `sphobjinv-textconv` need not be changed any further.

> A bit of basic documentation of how to set up the .inv textconv 
> No tutorials on anything git other than brief instructions for setting up the textconv. It's just beyond the scope for the sphobjinv 
> documentation.

On board with simple docs just saying for `git diff` support run optional command `sphobjinv --setup`. Then **not explaining the manual method** or the nitty gitty details.

> No direct testing of the operation of the git textconv.
> Nothing new **in the test suite** driving git or editing .gitattributes or .git/config.

Can move that out of fixtures and into code modules.

> So—if you're interested in paring down the PR to just the above

Yes would be interested. Would like to change the current PR rather than making a new PR.

Please confirm i can go ahead.

**Comment by msftcangoblowm** [comment]
> but, that magic command would end up needing to accommodate a wide variety of operating systems and user setups, and I'm 
> confident it would run into all sorts of hairy edge cases that would be a nightmare to debug and develop around.

When modifying `git/.config`, `sphobjinv-textconv` calls `git config`. So any hairy edge cases could only be caused by upstream.

When modifying `.gitattributes`, `sphobjinv-textconv` is doing that, but it's one line. The only edge case, is whether needs `\r\n` or not.

The setup changes are local; affecting one repo. Down the road, someone may request an option for these changes to be global, in `~/.config/git/config` and where ever the global `.gitattributes` is located.

There is a package specifically for platform user and site folders. The platform specific folders location will not an issue.

**Comment by bskinn** [comment]
What I'm really hesitant about is whether I want to have a `--setup` (or whatever name) flag in `sphobjinv` at all. (On rereading what I wrote, I wasn't very clear about this -- apologies there.)

On one hand, I actually just used `nbstripout` seriously for the first time this week, which has a similar automated `--install` functionality, and it _was_ really handy that it did it automatically for me.

On the other hand, these manipulations of git config and gitattributes really seem like something that should be encapsulated into their own library, and not rewritten in every library they're needed in. E.g., it seems like `nbstripout` has likely already implemented most or all of what would be required here ([example][nbstripout example])... and it would make way more sense to me, and I would be somewhat more inclined toward a `sphobjinv-textconv --setup` feature, if the git config/attributes manipulations were pulled into a library that `--setup` could call, instead of being re-implemented directly in the `sphobjinv` codebase. (Plus, from even a quick scan of the `nbstripout` code, I noticed multiple things that are sharp edges just waiting for users to eventually cut their fingers on... it's _hard_ to actually do this the right way.)

> Down the road, someone may request a local/global switch. 

Yeah... and I could see at least the following sorts of requests popping up, too:

- Can you make it install system-wide, not global or per-repo?
- Can you make it tell me where the textconv is installed?
- Can you make it uninstall the textconv for me?
- I have a complicated setup, and your `--setup` broke my git. Can you undo what you did?
- Can you make it install other gitconfig/gitattributes settings?
- Can you expose the `--setup` functionality in the API?

Just a really large possible CLI/API surface that might be requested.

----

Ultimately, this sort of git config/attributes manipulation is solidly outside `sphobjinv`'s mission, and I really don't want to open the door to issues/features surrounding the config/attributes manipulations themselves being on-topic for its issue tracker.

I'm open to continued discussion, but for now I'm only willing to consider the `sphobjinv-textconv` entrypoint itself, as an alias for `sphobjinv convert plain $0 -`, and documentation of the manual process for setting up the textconv.


[nbstripout example]: https://github.com/kynan/nbstripout/blob/main/nbstripout/_nbstripout.py#L197-L269

**Comment by msftcangoblowm** [comment]
You are right.

This should be a separate package, named git-filter, for interacting with git config and gitattributes.

Cuz it's not a sphobjinv specific issue. sphobjinv users could benefit from such a package.

```git-filter install --local --package=sphobjinv``
```git-filter remove --local --package=sphobjinv```
```git-filter list --package=nbsplitout```

And yaml files describing changes need by each package. Was thinking plugins, but that is overkill

**Comment by bskinn** [comment]
Thanks again for the work you did on this, @msftcangoblowm -- sorry it didn't end up being merged.

FYI, I did a bit of searching for other projects that have attempted this sort of git config, and found one: https://github.com/d12frosted/git-config-manager. I'm not sure what language it was written in, but it might be a useful resource.
